components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,374 @@
+
+        parseIdentityKeyStoreMappingConfigs();
+    }
+
+    public static IdentityKeyStoreResolver getInstance() {
+
+        if (instance == null) {
+            instance = new IdentityKeyStoreResolver();
+        }
+        return instance;
+    }
+
+    /**
+     * Return Primary or tenant keystore according to given tenant domain.
+     *
+     * @param tenantDomain  Tenant domain
+     * @return Primary or tenant keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public KeyStore getKeyStore(String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                // Get primary keystore from keyStoreManager
+                return keyStoreManager.getPrimaryKeyStore();
+            }
+
+            // Get tenant keystore from keyStoreManager
+            String tenantKeyStoreName = buildTenantKeyStoreName(tenantDomain);
+            return keyStoreManager.getKeyStore(tenantKeyStoreName);
+        } catch (Exception e) {
+            throw new IdentityKeyStoreResolverException(
+                    "Error occurred when retrieving keystore for tenant: " + tenantDomain, e);
+        }
+    }
+
+    /**
+     * Return Primary, tenant or custom keystore.
+     *
+     * @param tenantDomain      Tenant domain
+     * @param inboundProtocol   Inbound authentication protocol of the application
+     * @return Primary, tenant or custom keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public KeyStore getKeyStore(String tenantDomain, InboundProtocol inboundProtocol)
+            throws IdentityKeyStoreResolverException {
+
+        if (keyStoreMappings.containsKey(inboundProtocol)) {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain) ||
+                    keyStoreMappings.get(inboundProtocol).getUseInAllTenants()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Custom keystore configuration avialble for " + inboundProtocol + " protocol.");
+                }
+
+                String keyStoreName = buildCustomKeyStoreName(keyStoreMappings.get(inboundProtocol).getKeyStoreName());
+
+                try {
+                    int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                    KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+                    return keyStoreManager.getKeyStore(keyStoreName);
+                } catch (Exception e) {
+                    throw new IdentityKeyStoreResolverException(
+                            "Error occurred when retrieving keystore for protocol: " + inboundProtocol, e);
```

Comment:
Move Error messages to Constants + Add error codes

Reply:
Fixed in https://github.com/wso2/carbon-identity-framework/pull/5794/commits/4d420ef4b5e4f34ffb9eb0e792c349cf2bd13cf5